runtime/kubernetes/kubernetes.go

```diff
@@ -469,29 +252,58 @@ func (k *kubernetes) Update(s *runtime.Service, opts ...runtime.UpdateOption) er
 
 // Delete removes a service
 func (k *kubernetes) Delete(s *runtime.Service, opts ...runtime.DeleteOption) error {
+	k.Lock()
+	defer k.Unlock()
+
+	// parse the options
 	options := runtime.DeleteOptions{
 		Namespace: client.DefaultNamespace,
 	}
 	for _, o := range opts {
 		o(&options)
 	}
 
-	k.Lock()
-	defer k.Unlock()
-
-	// create new kubernetes micro service
-	service := newService(s, runtime.CreateOptions{
+	// delete the deployment
+	dep := client.NewDeployment(s, &runtime.CreateOptions{
 		Type:      k.options.Type,
 		Namespace: options.Namespace,
 	})
+	if err := k.client.Delete(dep, client.DeleteNamespace(options.Namespace)); err != nil {
+		if err == api.ErrNotFound {
+			return runtime.ErrNotFound
+		}
+		if logger.V(logger.DebugLevel, logger.DefaultLogger) {
+			logger.Debugf("Runtime failed to delete deployment: %v", err)
+		}
+		return err
+	}
 
-	// delete the service credentials
-	ns := client.DeleteNamespace(options.Namespace)
-	k.client.Delete(&client.Resource{Name: credentialsName(s), Kind: "secret"}, ns)
+	// delete the credentials
+	if err := k.deleteCredentials(s, &runtime.CreateOptions{Namespace: options.Namespace}); err != nil {
+		return err
+	}
 
-	if err := service.Stop(k.client, ns); err == api.ErrNotFound {
-		return runtime.ErrNotFound
-	} else if err != nil {
+	// if there are more deployments for this service, then don't delete it
+	labels := map[string]string{}
+	if len(s.Name) > 0 {
+		labels["name"] = client.Format(s.Name)
+	}
+
+	// get the existing services. todo: refactor to just get the deployments
+	services, err := k.getServices(client.GetNamespace(options.Namespace), client.GetLabels(labels))
+	if err != nil || len(services) > 0 {
+		return err
+	}
+
+	// delete the service
+	srv := client.NewService(s, &runtime.CreateOptions{
+		Type:      k.options.Type,
+		Namespace: options.Namespace,
+	})
+	if err := k.client.Delete(srv, client.DeleteNamespace(options.Namespace)); err != nil {
```

Comment:
In theory there's a tiny race condition here where someone is adding a new version at the same time someone else is deleting a version which would cause a delete of the k8s service when in fact there is a newly created micro service. small enough of a chance that I think we'll get away with it for now